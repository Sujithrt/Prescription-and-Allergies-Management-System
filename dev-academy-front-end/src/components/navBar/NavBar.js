import React, { useState } from 'react';
import AppBar from '@mui/material/AppBar';
import Box from '@mui/material/Box';
import Toolbar from '@mui/material/Toolbar';
import IconButton from '@mui/material/IconButton';
import Typography from '@mui/material/Typography';
import MenuItem from '@mui/material/MenuItem';
import Menu from '@mui/material/Menu';
import AccountCircle from '@mui/icons-material/AccountCircle';
import MoreIcon from '@mui/icons-material/MoreVert';
import { useHistory } from 'react-router-dom';
import './../navBar/navBar.css';
import SearchBar from './../searchBar/SearchBar';
import { navBarStyle } from './../../Constants'

/**
 * Renders the Navigation Bar for the React Application 
 */
function PrimarySearchAppBar() {

    const history = useHistory();

    const [anchorEl, setAnchorEl] = useState(null);
    const [mobileMoreAnchorEl, setMobileMoreAnchorEl] = useState(null);

    const isMenuOpen = Boolean(anchorEl);
    const isMobileMenuOpen = Boolean(mobileMoreAnchorEl);

    /**
     * Opens a drop down list upon clicking the Profile Icon
     * @param {onClick} event 
     */
    const handleProfileMenuOpen = (event) => {
        setAnchorEl(event.currentTarget);
    };

    /**
     * Handles Navigation bar menu close when app is being used on mobile
     */
    const handleMobileMenuClose = () => {
        setMobileMoreAnchorEl(null);
    };

    /**
     * Handles Navigation bar menu close 
     */
    const handleMenuClose = () => {
        setAnchorEl(null);
        handleMobileMenuClose();
    };

    /**
     * Handles Navigation bar menu open when app is being used on mobile
     * @param {onClick} event 
     */
    const handleMobileMenuOpen = (event) => {
        setMobileMoreAnchorEl(event.currentTarget);
    };

    /**
     * Opens User profile when clicked
     */
    const handleOpenProfile = () => {
        setAnchorEl(null);
        handleMobileMenuClose();
        history.push('/user');
    }

    /**
     * Logs out from the Application
     */
    const logOut = () => {
        setAnchorEl(null);
        handleMobileMenuClose();
        sessionStorage.clear();
        history.push('/');
        window.location.reload()
    }

    const menuId = 'primary-search-account-menu';

    /**
     * Render menu on Larger displays
     */
    const renderMenu = (
        <Menu
            anchorEl={anchorEl}
            anchorOrigin={{
                vertical: 'top',
                horizontal: 'right',
            }}
            id={menuId}
            keepMounted
            transformOrigin={{
                vertical: 'top',
                horizontal: 'right',
            }}
            data-testid="menu"
            open={isMenuOpen}
            onClose={handleMenuClose}
        >
            <MenuItem data-testid="openUserProfile" style={navBarStyle} onClick={handleOpenProfile}>Profile</MenuItem>
            <MenuItem data-testid="clickLogout" style={navBarStyle} onClick={logOut}>Logout</MenuItem>
        </Menu>
    );

    /**
     * Render menu on mobile or smaller displays
     */
    const mobileMenuId = 'primary-search-account-menu-mobile';
    const renderMobileMenu = (
        <Menu
            anchorEl={mobileMoreAnchorEl}
            anchorOrigin={{
                vertical: 'top',
                horizontal: 'right',
            }}
            id={mobileMenuId}
            keepMounted
            transformOrigin={{
                vertical: 'top',
                horizontal: 'right',
            }}
            open={isMobileMenuOpen}
            onClose={handleMobileMenuClose}
        >
            <MenuItem style={navBarStyle} onClick={handleOpenProfile}>Profile</MenuItem>
            <MenuItem style={navBarStyle} onClick={logOut}>Logout</MenuItem>
        </Menu>
    );

    return (
        <Box sx={{ flexGrow: 1 }}>
            <AppBar position="fixed">
                <Toolbar>
                    <Typography
                        variant="h6"
                        noWrap
                        component="div"
                        sx={{ display: { xs: 'none', sm: 'block' } }}
                    >
                        <div data-testid="pnamsHeader" onClick={()=> history.push('/dashboard')} className="nav-link" to="/dashboard">PNAMS</div>
                    </Typography>
                    <SearchBar />
                    <Box sx={{ flexGrow: 1 }} />
                    <Box sx={{ display: { xs: 'none', md: 'flex' } }}>
                        <IconButton
                            data-testid="profileMenu"
                            size="large"
                            edge="end"
                            aria-label="account of current user"
                            aria-controls={menuId}
                            aria-haspopup="true"
                            onClick={handleProfileMenuOpen}
                            color="inherit"
                        >
                            <AccountCircle />
                        </IconButton>
                    </Box>
                    <Box sx={{ display: { xs: 'flex', md: 'none' } }}>
                        <IconButton
                            data-testid="profileMenuMobile"
                            size="large"
                            aria-label="show more"
                            aria-controls={mobileMenuId}
                            aria-haspopup="true"
                            onClick={handleMobileMenuOpen}
                            color="inherit"
                        >
                            <MoreIcon />
                        </IconButton>
                    </Box>
                </Toolbar>
            </AppBar>
            {renderMobileMenu}
            {renderMenu}
        </Box>
    );
}

export default PrimarySearchAppBar