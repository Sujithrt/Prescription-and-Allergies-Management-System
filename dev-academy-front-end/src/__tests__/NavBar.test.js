import React from 'react';
import { shallowToJson } from "enzyme-to-json";
import { shallow, mount } from 'enzyme';
import { Router } from 'react-router-dom';
import NavBar from '../components/navBar/NavBar'

describe("NavBar Test", () => {
	it("NavBar is rendered as expected", () => {
		const navBar = shallow(<NavBar />);
		expect(shallowToJson(navBar)).toMatchSnapshot();
	});
	it('Check back to dashboard on App icon click', () => {
		const historyMock = { push: jest.fn(), location: {}, listen: jest.fn() };
		const navBar = mount(<Router history={historyMock}><NavBar /></Router>)
		navBar.find({ 'data-testid': 'pnamsHeader' }).simulate('click')
		expect(historyMock.push.mock.calls[0][0]).toEqual(`/dashboard`);
	});
});

describe("Profile Menu test", () => {
	it("test handleProfileMenuOpen", () => {
		const navBar = shallow(<NavBar />)
		navBar.find({ 'data-testid': 'profileMenu' }).simulate('click', { target: {} })
	});
	it("test handleMenuClose", () => {
		const navBar = shallow(<NavBar />)
		navBar.find({ 'data-testid': 'menu' }).simulate('close')
	});
	it("test handleProfileMenuOpen", () => {
		const navBar = shallow(<NavBar />)
		navBar.find({ 'data-testid': 'profileMenuMobile' }).simulate('click', { target: {} })
	});
});

describe("Menu Items test", () => {
	it("test handleOpenProfile", () => {
		const historyMock = { push: jest.fn(), location: {}, listen: jest.fn() };
		const navBar = mount(<Router history={historyMock}><NavBar /></Router>)
		navBar.find({ 'data-testid': 'openUserProfile' }).at(1).simulate('click')
		expect(historyMock.push.mock.calls[0][0]).toEqual(`/user`);
	});
	it("test handleMenuClose", () => {
		const historyMock = { push: jest.fn(), location: {}, listen: jest.fn() };
		const navBar = mount(<Router history={historyMock}><NavBar /></Router>)
		navBar.find({ 'data-testid': 'clickLogout' }).at(1).simulate('click')
		expect(historyMock.push.mock.calls[0][0]).toEqual(`/`);
	});
});