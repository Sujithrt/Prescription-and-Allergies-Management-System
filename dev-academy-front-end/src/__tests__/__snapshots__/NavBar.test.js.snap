// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar Test NavBar is rendered as expected 1`] = `
<ForwardRef(Box)
  sx={
    Object {
      "flexGrow": 1,
    }
  }
>
  <ForwardRef(AppBar)
    position="fixed"
  >
    <ForwardRef(Toolbar)>
      <ForwardRef(Typography)
        component="div"
        noWrap={true}
        sx={
          Object {
            "display": Object {
              "sm": "block",
              "xs": "none",
            },
          }
        }
        variant="h6"
      >
        <div
          className="nav-link"
          data-testid="pnamsHeader"
          onClick={[Function]}
          to="/dashboard"
        >
          PNAMS
        </div>
      </ForwardRef(Typography)>
      <SearchBar />
      <ForwardRef(Box)
        sx={
          Object {
            "flexGrow": 1,
          }
        }
      />
      <ForwardRef(Box)
        sx={
          Object {
            "display": Object {
              "md": "flex",
              "xs": "none",
            },
          }
        }
      >
        <ForwardRef(IconButton)
          aria-controls="primary-search-account-menu"
          aria-haspopup="true"
          aria-label="account of current user"
          color="inherit"
          data-testid="profileMenu"
          edge="end"
          onClick={[Function]}
          size="large"
        >
          <Memo(ForwardRef(AccountCircleIcon)) />
        </ForwardRef(IconButton)>
      </ForwardRef(Box)>
      <ForwardRef(Box)
        sx={
          Object {
            "display": Object {
              "md": "none",
              "xs": "flex",
            },
          }
        }
      >
        <ForwardRef(IconButton)
          aria-controls="primary-search-account-menu-mobile"
          aria-haspopup="true"
          aria-label="show more"
          color="inherit"
          data-testid="profileMenuMobile"
          onClick={[Function]}
          size="large"
        >
          <Memo(ForwardRef(MoreVertIcon)) />
        </ForwardRef(IconButton)>
      </ForwardRef(Box)>
    </ForwardRef(Toolbar)>
  </ForwardRef(AppBar)>
  <ForwardRef(Menu)
    anchorEl={null}
    anchorOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
    id="primary-search-account-menu-mobile"
    keepMounted={true}
    onClose={[Function]}
    open={false}
    transformOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
  >
    <ForwardRef(MenuItem)
      onClick={[Function]}
      style={
        Object {
          "WebkitBoxAlign": "center",
          "WebkitBoxPack": "start",
          "alignItems": "center",
          "border": 0,
          "borderRadius": 0,
          "color": "inherit",
          "cursor": "pointer",
          "display": "flex",
          "fontFamily": "\\"Roboto\\",\\"Helvetica\\",\\"Arial\\",sans-serif",
          "fontSize": "1rem",
          "fontWeight": 400,
          "justifyContent": "flex-start",
          "letterSpacing": "0.00938em",
          "lineHeight": 1.5,
          "margin": 0,
          "outline": 0,
          "padding": "5px 10px",
          "position": "relative",
          "userSelect": "none",
          "verticalAlign": "middle",
        }
      }
    >
      Profile
    </ForwardRef(MenuItem)>
    <ForwardRef(MenuItem)
      onClick={[Function]}
      style={
        Object {
          "WebkitBoxAlign": "center",
          "WebkitBoxPack": "start",
          "alignItems": "center",
          "border": 0,
          "borderRadius": 0,
          "color": "inherit",
          "cursor": "pointer",
          "display": "flex",
          "fontFamily": "\\"Roboto\\",\\"Helvetica\\",\\"Arial\\",sans-serif",
          "fontSize": "1rem",
          "fontWeight": 400,
          "justifyContent": "flex-start",
          "letterSpacing": "0.00938em",
          "lineHeight": 1.5,
          "margin": 0,
          "outline": 0,
          "padding": "5px 10px",
          "position": "relative",
          "userSelect": "none",
          "verticalAlign": "middle",
        }
      }
    >
      Logout
    </ForwardRef(MenuItem)>
  </ForwardRef(Menu)>
  <ForwardRef(Menu)
    anchorEl={null}
    anchorOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
    data-testid="menu"
    id="primary-search-account-menu"
    keepMounted={true}
    onClose={[Function]}
    open={false}
    transformOrigin={
      Object {
        "horizontal": "right",
        "vertical": "top",
      }
    }
  >
    <ForwardRef(MenuItem)
      data-testid="openUserProfile"
      onClick={[Function]}
      style={
        Object {
          "WebkitBoxAlign": "center",
          "WebkitBoxPack": "start",
          "alignItems": "center",
          "border": 0,
          "borderRadius": 0,
          "color": "inherit",
          "cursor": "pointer",
          "display": "flex",
          "fontFamily": "\\"Roboto\\",\\"Helvetica\\",\\"Arial\\",sans-serif",
          "fontSize": "1rem",
          "fontWeight": 400,
          "justifyContent": "flex-start",
          "letterSpacing": "0.00938em",
          "lineHeight": 1.5,
          "margin": 0,
          "outline": 0,
          "padding": "5px 10px",
          "position": "relative",
          "userSelect": "none",
          "verticalAlign": "middle",
        }
      }
    >
      Profile
    </ForwardRef(MenuItem)>
    <ForwardRef(MenuItem)
      data-testid="clickLogout"
      onClick={[Function]}
      style={
        Object {
          "WebkitBoxAlign": "center",
          "WebkitBoxPack": "start",
          "alignItems": "center",
          "border": 0,
          "borderRadius": 0,
          "color": "inherit",
          "cursor": "pointer",
          "display": "flex",
          "fontFamily": "\\"Roboto\\",\\"Helvetica\\",\\"Arial\\",sans-serif",
          "fontSize": "1rem",
          "fontWeight": 400,
          "justifyContent": "flex-start",
          "letterSpacing": "0.00938em",
          "lineHeight": 1.5,
          "margin": 0,
          "outline": 0,
          "padding": "5px 10px",
          "position": "relative",
          "userSelect": "none",
          "verticalAlign": "middle",
        }
      }
    >
      Logout
    </ForwardRef(MenuItem)>
  </ForwardRef(Menu)>
</ForwardRef(Box)>
`;
